modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -90,276 +92,256 @@ interface WalkResult {
+  purify.addHook('uponSanitizeElement', (ele, evt) => {
+    const element = SugarElement.fromDom(ele);
+    if (!SugarNode.isElement(element) || SugarNode.name(element) === 'body') {
+      return;
+    }
+    const rule = schema.getElementRule(evt.tagName.toLowerCase());
+    if (!rule) {
+      Remove.unwrap(element);
+      return;
+    }
+    if (!Obj.has(evt.allowedTags, evt.tagName)) {
+      evt.allowedTags[evt.tagName] = true;
+    }
 
-    for (let textNode = node.prev; textNode && textNode.type === 3;) {
-      const textVal = textNode.value.replace(endWhiteSpaceRegExp, '');
+    const bogus = !Attribute.has(element, 'data-mce-type') && Attribute.get(element, 'data-mce-bogus');
+    if (bogus === 'all') {
+      Remove.remove(element);
+      return;
+    } else if (bogus === '1') {
+      Remove.unwrap(element);
+      return;
+    }
 
-      // Found a text node with non whitespace then trim that and break
-      if (textVal.length > 0) {
-        textNode.value = textVal;
+    // Fix the attributes for the element, unwrapping it if we have to
+    Arr.each(rule.attributesForced ?? [], (attr) =>
+      Attribute.set(element, attr.name, attr.value === '{$uid}' ? `mce_${uid++}` : attr.value)
+    );
+    Arr.each(rule.attributesDefault ?? [], (attr) => {
+      if (!Attribute.has(element, attr.name)) {
+        Attribute.set(element, attr.name, attr.value === '{$uid}' ? `mce_${uid++}` : attr.value);
+      }
+    });
+    if (rule.attributesRequired) {
+      if (!Arr.exists(rule.attributesRequired, (attr) => Attribute.has(element, attr))) {
+        Remove.unwrap(element);
         return;
       }
+    }
+    if (rule.removeEmptyAttrs && Attribute.hasNone(element)) {
+      Remove.unwrap(element);
+      return;
+    }
 
-      const textNodeNext = textNode.next;
+    if (rule.outputName && rule.outputName !== SugarNode.name(element)) {
+      const replacement = SugarElement.fromTag(rule.outputName);
 
-      // Fix for bug #7543 where bogus nodes would produce empty
-      // text nodes and these would be removed if a nested list was before it
-      if (textNodeNext) {
-        if (textNodeNext.type === 3 && textNodeNext.value.length) {
-          textNode = textNode.prev;
-          continue;
-        }
+      const allAttributes = Arr.map(element.dom.attributes, (attr) => attr.name);
+      Attribute.transfer(element, replacement, allAttributes);
 
-        if (!blockElements[textNodeNext.name] && textNodeNext.name !== 'script' && textNodeNext.name !== 'style') {
-          textNode = textNode.prev;
-          continue;
-        }
-      }
+      const children = Traverse.children(element);
+      InsertAll.append(replacement, children);
 
-      const sibling = textNode.prev;
-      textNode.remove();
-      textNode = sibling;
+      Insert.after(element, replacement);
+      Remove.remove(element);
```

Comment:
I mentioned this via DM, but copying here as well... I believe this can be changed to this if you change mutate to insert after as well:
```
      const replacement = Replication.mutate(SugarElement.fromDom(element), rule.outputName);
```